src/treeview/selection.js

```diff
@@ -0,0 +1,330 @@
+		 */
+		this._lastRangeBackward = false;
+	}
+
+	/**
+	 * Selection anchor. Anchor may be described as a position where the selection starts. Together with
+	 * {@link engine.treeView.Selection#focus focus} they define the direction of selection, which is important
+	 * when expanding/shrinking selection. Anchor is always the start or end of the most recent added range.
+	 * It may be a bit unintuitive when there are multiple ranges in selection.
+	 *
+	 * @see engine.treeView.Selection#focus
+	 * @type {engine.treeView.Position}
+	 */
+	get anchor() {
+		if ( !this._ranges.length ) {
+			return null;
+		}
+		const range = this._ranges[ this._ranges.length - 1 ];
+		const anchor = this._lastRangeBackward ? range.end : range.start;
+
+		return Position.createFromPosition( anchor );
+	}
+
+	/**
+	 * Selection focus. Focus is a position where the selection ends.
+	 *
+	 * @see engine.treeView.Selection#anchor
+	 * @type {engine.treeView.Position}
+	 */
+	get focus() {
+		if ( !this._ranges.length ) {
+			return null;
+		}
+		const range = this._ranges[ this._ranges.length - 1 ];
+		const focus = this._lastRangeBackward ? range.start : range.end;
+
+		return Position.createFromPosition( focus );
+	}
+
+	/**
+	 * Returns whether the selection is collapsed. Selection is collapsed when all it's ranges are collapsed.
+	 *
+	 * @type {Boolean}
+	 */
+	get isCollapsed() {
+		for ( let range of this._ranges ) {
+			if ( !range.isCollapsed ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Returns number of ranges in selection.
+	 *
+	 * @type {Number}
+     */
+	get rangeCount() {
+		return this._ranges.length;
+	}
+
+	/**
+	 * Adds a range to the selection. Added range is copied. This means that passed range is not saved in the
+	 * selection instance and you can safely operate on it.
+	 *
+	 * Accepts a flag describing in which way the selection is made - passed range might be selected from
+	 * {@link engine.treeView.Range#start start} to {@link engine.treeView.Range#end end}
+	 * or from {@link engine.treeView.Range#end end} to {@link engine.treeView.Range#start start}.
+	 * The flag is used to set {@link engine.treeView.Selection#anchor anchor} and
+	 * {@link engine.treeView.Selection#focus focus} properties.
+	 *
+	 * Throws {@link utils.CKEditorError CKEditorError} `view-selection-range-intersects` if added range intersects
+	 * with ranges already stored in Selection instance.
+	 *
+	 * @fires engine.treeView.Selection#change:range
+	 * @param {engine.treeView.Range} range
+	 */
+	addRange( range, isBackward ) {
+		this._pushRange( range );
+		this._lastRangeBackward = !!isBackward;
+		this.fire( 'change:range' );
+	}
+
+	/**
+	 * Returns an array of ranges added to the selection. The method returns a copy of internal array, so
+	 * it will not change when ranges get added or removed from selection.
+	 *
+	 * @returns {Array.<engine.treeView.Range>}
+	 */
+	getRanges() {
+		return this._ranges.slice();
```

Comment:
It could be a generator: `return this. _ranges[ Symbol.iterator ]();` See, for instance: https://github.com/ckeditor/ckeditor5-engine/blob/master/src/treemodel/node.js#L176-L178